capsules/src/symmetric_encryption.rs

```diff
@@ -137,9 +137,45 @@ impl<'a, E: SymmetricEncryptionDriver + 'a> Crypto<'a, E> {
             state: Cell::new(CryptoState::IDLE),
         }
     }
+
+    fn aes128_crypt_ctr(&self, state: CryptoState) -> ReturnCode {
+        if self.key_configured.get() && !self.busy.get() && self.state.get() == CryptoState::IDLE {
+            for cntr in self.apps.iter() {
+                cntr.enter(|app, _| {
+                    self.busy.set(true);
+                    self.state.set(state);
+                    app.data_buf.as_ref().map(|slice| {
+                        let len1 = slice.len();
+                        // Copy data from app slice to kernel buffer.
+                        self.kernel_data.take().map(|buf| {
+                            for (out, inp) in buf.iter_mut().zip(slice.as_ref()[0..len1].iter()) {
+                                *out = *inp;
+                            }
+                            app.ctr_buf.as_ref().map(|slice2| {
+                                let len2 = slice2.len();
+                                // Copy counter value to kernel buffer.
+                                self.kernel_ctr.take().map(move |ctr| {
+                                    for (out, inp) in ctr.iter_mut()
+                                        .zip(slice2.as_ref()[0..len2].iter()) {
+                                        *out = *inp;
+                                    }
+                                    self.crypto.aes128_crypt_ctr(buf, ctr, len1);
```

Comment:
Not really... what would go wrong if the input buffer were shorter than 16 bytes? the `zip` operation ensures we won't go beyond the smaller of `out.len()` or `slice2[0..len2].len()`. Is the problem that there may be bytes remaining at the end of the counter from a previous call? In which case, would it be fine to just pad `inp` with zeroes if it's not big enough?